backend/zkppSalt.js

```diff
@@ -0,0 +1,229 @@
+import { blake32Hash } from '~/shared/functions.js'
+import { timingSafeEqual } from 'crypto'
+import nacl from 'tweetnacl'
+import sbp from '@sbp/sbp'
+import { boxKeyPair, encryptContractSalt, hashStringArray, hashRawStringArray, hash, parseRegisterSalt, randomNonce, computeCAndHc } from '~/shared/zkpp.js'
+
+// TODO HARDCODED VALUES
+const recordPepper = 'pepper'
+const recordMasterKey = 'masterKey'
+const challengeSecret = 'secret'
+const registrationSecret = 'secret'
+const maxAge = 30
+
+const getZkppSaltRecord = async (contract: string) => {
+  const recordId = blake32Hash(hashStringArray('RID', contract, recordPepper))
```

Comment:
A double hash is not required, but is here as all other 'records' use blake32.

Reply:
@corrideat In that case, could you redefine `hashStringArray` to use blake32 instead of `nacl.hash`?

Reply:
Any reason for this? SHA-512 is generally a more widely-used hash than Blake32, and the main advantage here (besides limiting the surface of the 'crypto' code to NaCl) is also the API. nacl.hash returns a Uint8Array, which makes it convenient for some of the key derivation uses in the code, whereas the blake32 code returns a string, which would then need to be converted back to a buffer.

Reply:
Well, we're designing a protocol, right? What's happening above seems to have occurred because of a quirk of the various libraries we're using, but at the end of the day we're supposed to be implementing a protocol. And a protocol is supposed to be clean and elegant from the get-go and inform the implementation, not the other way around. So that's what I'm trying to say - how is this code supposed to look if it were the implementation of a clean protocol?

Reply:
It is a decision partly influenced by the libraries used, yes, although I'm not so sure that this makes the protocol any more or less clean. Blake32 is used primarily for deriving database entry keys (talking in general, not in this particular instance), and, while there's no reason why Blake32 must be used in this particular case, it helps keep things consistent.

Whether Blake32 or something else entirely is used plays no role in the protocol, because clients are not supposed to interact with these records directly. In fact, they don't even need to be stored publicly, if we were to use a separate server database.

Now, for the protocol itself, which is where `hashStringArray` is used, in principle we could use any hash function. We're already using NaCl for all of the crypto code (and scrypt), including SHA-512 by proxy (as that's what NaCl is using internally). AFAIK there's nothing wrong with SHA-512 as a hash function, and apart from having a nice API that greatly simplifies the code (i.e., using bytes instead of strings), it reduces the amount of crypto code that we need in crypto implementations.

Reply:
OK, that sounds reasonable, thanks for the explanation @corrideat!